_posts/2019-05-23-quarkus-and-angular-ui-development-mode.adoc

```diff
@@ -0,0 +1,391 @@
+        }
+        if (path.startsWith("/api/") || path.startsWith("/servlet/")) {
+            return false;
+        }
+
+        return true;
+    }
+}
+----
+
+All this does is to check the path that is used. If it looks like it is a file, or starts with `/api/` or `/servlet/` we don't do anything special. However, if it is something else, we assume it is an Angular route, and use the request dispatcher to forward the request. This is why in Step 3 we decided to put our REST endpoints and servlets under a specific path to make this decision easier. Still, this is fully customisable in your applications, so it is up to you how to handle this yourself.
+
+To try it out, package and start the application by running:
+[source,bash]
+----
+$mvn package quarkus:dev -Dui.dev
+----
+If you go to http://localhost:8080 you will see the initial page again. This time all the links work and we can refresh on any page we want! This is progress, and we now have a fully working application.
+
+However, we still need to restart and repackage our application every time we want to change something in the UI. The next two steps will show how to make this more convenient. Stop Quarkus before going to the next step.
+
+==== Step 5 - Angular proxy for dual development mode
+The diff for this step can be found https://github.com/kabir/blog-quarkus-ui-development/compare/step-4..step-5[here].
+
+Angular CLI ships with a https://angular.io/guide/build#proxying-to-a-backend-server[proxy]. Although I am not familiar with these other frameworks, from a quick search it seems that https://facebook.github.io/create-react-app/docs/proxying-api-requests-in-development[React] and https://cli.vuejs.org/config/#devserver-proxy[Vue] have something similar.
+
+The changes are simple.
+
+First we create a `webapp/proxy.conf.json`:
+[source,json]
+----
+{
+    "/api/*": {
+        "target": "http://localhost:8080",
+        "secure": false
+    }
+}
+----
+This basically tells Angular that when making REST calls where the path starts with `/api/` we should direct to the back-end server running on port `8080`. This is basically the application running in Quarkus. Angular CLI itself runs on port `4200`.
+
+The next thing we need to do is to add another `script` entry to `package.json`:
+[source,json]
+----
+"scripts": {
+    "ng": "ng",
+    "start": "ng serve",
+    "proxy": "ng serve --proxy-config proxy.conf.json",
+    ...
+----
+
+Now if we start Angular with `yarn proxy` (rather than the standard `ng serve`) it will use the proxy configuration we just set up.
+
+Now let's try it out. You need two terminals.
+
+In the first terminal run:
+[source,bash]
+----
+$mvn clean
+$mvn package quarkus:dev
+----
+to start the Quarkus application. Note how we did not pass in `-Dui.dev` so we will no longer bundle the web application which saves us a significant amount of time. If you go to http://localhost:8080 you will get an error since we have no web application this time.
```

Comment:
`you will get an error`

That's true if you remove `webapp/dist/webapp` first.